pkg/kn/traffic/compute.go

```diff
@@ -0,0 +1,362 @@
+			if target.Tag != "" { // referenced revision is requested to have multiple tags
+				break
+			} else {
+				e[i].Tag = tag // referenced revision doesn't have tag, tag it
+				return e
+			}
+		}
+	}
+	// append a new target if revision doesn't exist in traffic block
+	// or if referenced revision is requested to have multiple tags
+	e = append(e, newTarget(tag, revision, 0, false))
+	return e
+}
+
+// TagLatestRevision assigns given tag to latest ready revision
+func (e ServiceTraffic) TagLatestRevision(tag string) ServiceTraffic {
+	for i, target := range e {
+		if *target.LatestRevision {
+			e[i].Tag = tag
+			return e
+		}
+	}
+	e = append(e, newTarget(tag, "", 0, true))
+	return e
+}
+
+// SetTrafficByRevision checks given revision in existing traffic block and sets given percent if found
+func (e ServiceTraffic) SetTrafficByRevision(revision string, percent int) {
+	for i, target := range e {
+		if target.RevisionName == revision {
+			e[i].Percent = percent
+			break
+		}
+	}
+}
+
+// SetTrafficByTag checks given tag in existing traffic block and sets given percent if found
+func (e ServiceTraffic) SetTrafficByTag(tag string, percent int) {
+	for i, target := range e {
+		if target.Tag == tag {
+			e[i].Percent = percent
+			break
+		}
+	}
+}
+
+// SetTrafficByLatestRevision sets given percent to latest ready revision of service
+func (e ServiceTraffic) SetTrafficByLatestRevision(percent int) {
+	for i, target := range e {
+		if *target.LatestRevision {
+			e[i].Percent = percent
+			break
+		}
+	}
+}
+
+// ResetAllTargetPercent resets (0) 'Percent' field for all the traffic targets
+func (e ServiceTraffic) ResetAllTargetPercent() {
+	for i := range e {
+		e[i].Percent = 0
+	}
+}
+
+// RemoveNullTargets removes targets from traffic block if they don't have and 0 percent traffic
+func (e ServiceTraffic) RemoveNullTargets() (newTraffic ServiceTraffic) {
+	for _, target := range e {
+		if target.Tag == "" && target.Percent == 0 {
+		} else {
+			newTraffic = append(newTraffic, target)
+		}
+	}
+	return newTraffic
+}
+
+func errorOverWritingtagOfLatestReadyRevision(existingTag, requestedTag string) error {
+	return fmt.Errorf("tag '%s' exists on latest ready revision of service, "+
+		"refusing to overwrite existing tag with '%s', "+
+		"add flag '--untag %s' in command to untag it", existingTag, requestedTag, existingTag)
+}
+func errorOverWritingTag(tag string) error {
+	return fmt.Errorf("refusing to overwrite existing tag in service, "+
+		"add flag '--untag %s' in command to untag it", tag)
+}
+
+func errorRepeatingLatestRevision(forFlag string) error {
+	return fmt.Errorf("repetition of identifier %s "+
+		"is not allowed, use only once with %s flag", latestRevisionRef, forFlag)
+}
+
+// verifies if user has repeated @latest field in --tag or --traffic flags
+// verifyInputSanity checks:
+// - if user has repeated @latest field in --tag or --traffic flags
+// - if provided traffic portion are integers
```

Comment:
Hi @navidshaikh 

I think except `@latest`, we should also report error if there're duplicated revision name in the traffic flags, for e.g. so far we will fail util the server side, we can check this from the client side. I submitted a PR to your branch and please take a review. Thanks!

```
./kn service update canary --traffic gracecanary-blue=40 --traffic gracecanary-blue=60
Internal error occurred: admission webhook "webhook.serving.knative.dev" denied the request: mutation failed: Traffic targets sum to 50, want 100: spec.traffic
```

Reply:
Hi @zhanggbj , sounds good. 
If we are to add this check, I'd say, repetition of revision reference in `--traffic` block is not allowed (than saying revision name), because we allow `revisionName`, `tag` and `@latest` with `--traffic` flag.